src/python/pants/engine/rules.py

```diff
@@ -413,21 +422,36 @@ class Rule(ABC):
     def output_type(self):
         """An output `type` for the rule."""
 
-    @property
-    @abstractmethod
-    def dependency_rules(self):
-        """A tuple of @rules that are known to be necessary to run this rule.
 
-        Note that installing @rules as flat lists is generally preferable, as Rules already
-        implicitly form a loosely coupled RuleGraph: this facility exists only to assist with
-        boilerplate removal.
-        """
+def register_rules(*namespaces: Union[ModuleType, Mapping[str, Any]]) -> Iterable[Rule]:
```

Comment:
It's unlikely to be the slowest part of startup, but none of our calls to `def rules` are memoized. Perhaps the inner portion of this method (post module identification) could be.

Reply:
Yeah. The fact that we re-export rules at all / need to seems problematic. I have not thought through why we need to do that which is what leads to the need for this optimization now IIUC. It would be great to kill that need instead / in addition.

Reply:
Mm. There was some discussion of this the other day: https://pantsbuild.slack.com/archives/C0D7TNJHL/p1595565944292800 ... but yea, unclear what to do, as it does seem to make things easier in some cases.